src/fido2_child/fido2_child_common.c

```diff
@@ -0,0 +1,310 @@
+            data->user_verification = FIDO_OPT_FALSE;
+        } else if (user_verification != NULL) {
+            ERROR("[%s] is not a valid user-verification value.\n",
+                  user_verification);
+            ret = EINVAL;
+            goto done;
+        }
+    }
+
+    if (type != NULL) {
+        ret = cose_str_to_int(type, &data->type);
+        if (ret != EOK) {
+            ERROR("[%s] is not a valid COSE type (es256, rs256 or eddsa).\n",
+                  type);
+            goto done;
+        }
+    }
+
+    debug_prg_name = talloc_asprintf(NULL, "fido2_child[%d]", getpid());
+    if (debug_prg_name == NULL) {
+        ERROR("talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    if (debug_fd != -1) {
+        opt_logger = sss_logger_str[FILES_LOGGER];
+        ret = set_debug_file_from_fd(debug_fd);
+        if (ret != EOK) {
+            opt_logger = sss_logger_str[STDERR_LOGGER];
+            ERROR("set_debug_file_from_fd failed.\n");
+        }
+    }
+
+    DEBUG_INIT(debug_level, opt_logger);
+
+    ret = EOK;
+
+done:
+    return ret;
+}
+
+errno_t
+check_arguments(const struct fido2_data *data)
+{
+    errno_t ret = EOK;
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Argument values after parsing\n");
+    DEBUG(SSSDBG_TRACE_FUNC, "action: %d\n", data->action);
+    DEBUG(SSSDBG_TRACE_FUNC, "shortname: %s, domain: %s\n",
+          data->shortname, data->domain);
+    DEBUG(SSSDBG_TRACE_FUNC, "public_key: %s, key_handle: %s\n",
+          data->public_key, data->key_handle);
+    DEBUG(SSSDBG_TRACE_FUNC, "type: %d\n", data->type);
+    DEBUG(SSSDBG_TRACE_FUNC, "user_verification: %d\n",
+          data->user_verification);
+    DEBUG(SSSDBG_TRACE_FUNC, "debug_libfido2: %d\n", data->debug_libfido2);
+
+    if (data->action == ACTION_NONE) {
+        DEBUG(SSSDBG_OP_FAILURE, "No action set.\n");
+        ret = ERR_INPUT_PARSE;
+        goto done;
+    }
+
+    if (data->action == ACTION_REGISTER
+        && (data->shortname == NULL || data->domain == NULL)) {
+        DEBUG(SSSDBG_OP_FAILURE, "Too few arguments for register action.\n");
+        ret = ERR_INPUT_PARSE;
+        goto done;
+    }
+
+    if (data->action == ACTION_AUTHENTICATE
+        && (data->shortname == NULL || data->domain == NULL
+        || data->public_key == NULL || data->key_handle == NULL)) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Too few arguments for authenticate action.\n");
+        ret = ERR_INPUT_PARSE;
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+errno_t
+register_key(struct fido2_data *data)
+{
+    fido_cred_t *cred = NULL;
+    fido_dev_t *dev = NULL;
+    fido_dev_info_t *dev_list = NULL;
+    size_t dev_number = 0;
+    errno_t ret;
+
+    cred = fido_cred_new();
+    if (cred == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_new failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = prepare_credentials(data, cred);
```

Comment:
Hi,

would it be possible to call `prepare_credentials()` after the device is selected? In this case it would be possible to call `fido_cred_set_uv()` in `prepare_credentials() as well.

bye,
Sumit

Reply:
Yes, it's possible and I've changed the order in the last commit